### ISSUE TITLE
QPDF fails to compile in C++20 mode due to incomplete type since vector's destructor is now constexpr

### ISSUE THREAD

**Issue by zclifford** [qpdf/qpdf#942]
I realize this may be a bit premature, but I would like to be able to compile QPDF under C++20 mode without local patches.

C++20 makes more standard library types like std::vector<T> constexpr than previously.

This can cause compilation errors when T is an incomplete type when when some code using the vector is compiled. I found some background discussion about the issue [here](https://discourse.llvm.org/t/proposal-adding-a-c-20-early-adopter-mode-for-libc-constexpr-annotations-in-c-17/66446) if you'd like further details.

---

W.R.T. QPDF I believe this is only a problem in one spot in the file include/qpdf/JSON.hh:

```
class JSON
{
    ...
    struct JSON_array: public JSON_value
    {
        JSON_array() :
            JSON_value(vt_array)
        {
        }
        virtual ~JSON_array() = default;
        virtual void write(Pipeline*, size_t depth) const;
        std::vector<JSON> elements;
    };
   ...
};
```

Here JSON is an incomplete type when the (constexpr) std::vector destructor is compiled, triggering a compilation error on certain C++20 compilers, which throw an ugly message similar to:

```
include/c++/v1/vector:839:62: error: arithmetic on a pointer to an incomplete type 'JSON'
        __alloc_traits::destroy(__alloc(), std::__to_address(--__soon_to_be_end));
...
JSON.hh:383:17: note: in instantiation of member function 'std::vector<JSON>::~vector' requested here
        virtual ~JSON_array() = default;
```

---

For the moment I workaround this with a local patch which forward declares JSON_array:

```
class JSON
{
    ...
    struct JSON_array; 
    ...
};
struct JSON::JSON_array : public JSON_value 
{
    ...
};
```

---

Here is my best shot at a reduced testcase:

```
#include <vector>

struct JSON {
  struct JSON_value {
    virtual ~JSON_value() = default;
  };
  struct JSON_array : JSON_value {
    virtual ~JSON_array() = default;
    std::vector<JSON> elements;
  };
};
```

Note that it can be challenging to trigger the compilation error since it requires both a sufficiently new compiler and library version.  So let me know if you want more details reproduction steps. 

With my reduced testcase I can reproduce the issue with godbolt clag and `-std=c++20` flag. Or on my work (gLinux) system with `clang++ -v main.cc -std=c++20 -stdlib=libstdc++`.

However I wasn't able to easily reproduce the issue on my personal Debian computer. Presumably I either misconfigured something or the compiler or standard library wasn't new enough to tickle the problem.

**Comment by jberkenbilt** [comment]
I really looking forward to being able to use C++-20. There are a few enhancements that will be much easier with with. While I'm not close to being ready to require C++-20 yet (we just bumped to 17), I would accept a PR that allows the code to work with C++-20 while not breaking C++-17 support (or anything else).

**Comment by jberkenbilt** [comment]
Fixed by #966 which is the same change as #943 .
